shoop/campaigns/modules.py

```diff
@@ -111,21 +111,21 @@ def get_discount_line(campaign, amount, price_so_far):
         best_discount_campaign = None
         lines = []
         for campaign in matching_campaigns:
-            effect = campaign.effects.first()
-            if not effect:
-                continue
-
-            if hasattr(effect, "get_discount_lines"):
-                continue
-
-            discount_amount = effect.apply_for_basket(order_source=order_source)
-            # if campaign has coupon, match it to order_source.codes
-            if campaign.coupon:
-                # campaign was found because discount code matched. This line is always added
-                lines.append(get_discount_line(campaign, discount_amount, price_so_far))
-            elif best_discount is None or discount_amount > best_discount:
-                best_discount = discount_amount
-                best_discount_campaign = campaign
+            for effect in campaign.effects.all():
+                if not effect:
+                    continue
+
+                if hasattr(effect, "get_discount_lines"):
+                    continue
```

Comment:
Why are you mixing two kinds of effects in the same list? If the BasketLineEffects and BasketDiscountEffects were using two different related names, then there would be no need for this hasattr hack.